rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,126 @@
+
+### Random ops
+
+Legacy random ops, such as `tf.random.normal`, are not deterministic if no seed is set, and so such ops will raise a `FailedPreconditionError` when determinism is enabled. To fix, the user should set a global seed with `tf.random.set_seed`. Since most models use legacy random ops (for variable initialization and various other uses), in practice users must call `tf.random.set_seed` when enabling deterministic ops. Alternatively, users can pass a seed to every individual random operation, but doing so is more inconvenient.
+
+Certain random ops, such as `tf.image.sample_distorted_bounding_box` and `tf.nn.fractional_max_pool`, ignore the global seed if a seed is not explicitly passed. For such ops, setting the global seed is not enough to avoid the error, so users must pass a seed directly to the op.
+
+As for TensorFlow 2 random number generation, `tf.random.Generator.from_non_deterministic_state` will raise an error if called when determinism is enabled. In such cases, users should check if determinism is enabled and if so, use a different generator from a deterministic source. `tf.random.get_global_generator` implicitly calls `from_non_deterministic_state` if no global generator is set, and so will also raise an error if a global generator is not set with `tf.random.set_global_generator`.
+
+Stateless random functions, such as `tf.random.stateless_normal`, are always deterministic and so will never raise determinism-related errors.
+
+No error will be raised if a random op or generator is run before determinism is enabled (as is true for any other op), so users should take care to enable determinism before running any random ops or generators.
+
+### Testing plan
+
+We must ensure that every op will either run deterministically or raise an error if `enable_deterministic_ops` has been called. In order to do this, we must test for determinism, both to check we don't accidentally miss any nondeterministic ops and to ensure any previously-deterministic ops are not accidentally made nondeterministic in the future. To accomplish this, we will have the following tests:
+
+1. We will add tests to several of the [official models](https://github.com/tensorflow/models/tree/master/official) to ensure they run deterministically. In particular, each test will train a model for several steps, then retrain it from scratch several times. The final weights after training will be asserted to be the same each time. This tests not only the `enable_deterministic_ops` API but that the entire model is deterministic. This only tests ops that the official models use.
+
+2. We will add a special mode to TensorFlow where every time a non-stateful op is run, TensorFlow will rerun the op several times and assert the outputs are the same each time. We will then run the TensorFlow unit tests with this mode as part of the nightly tests. Doing so ensures that for each op that is run as part of a unit test, it will be tested for determinism.
```

Comment:
Most forward tests are too small and/or not designed in some other way to test reproducibility. For example, many tests use integers even when exercising floating-point data-paths. To make a test that is likely to exercises nondeterminism in a forward-path, it needs to use the op in a natural way, which most unit tests don't, or don't do much of. Most test cases are too small to exercise the king of nondeterminism we want to protect against, which shows up with test cases that span asynchronous compute engines.

For the backwards paths, the way that unit tests are written, comparing analytical and numerical Jacobian matrices will not catch nondeterminism. This is because the analytical Jacobian matrix does not capture the backprop effect on real upstream gradients. The existing test literally cannot see the nondeterminism. This is one of the factors that motivated me to develop the [gradient injection](https://github.com/NVIDIA/framework-determinism/blob/master/test/gradient_injection.md) approach to testing backprop determinism. Apart from that, even if the analytical Jacobians could capture the nondeterminism, the test cases would have to be made prohibitively large. The traditional method of gradient function testing, using the Jacobians, requires that the test cases be necessarily small (because the size of the Jacobians explodes in O(Nin*Nout), where N is the number of elements).

Reply:
Automatic generation of effective determinism tests is a hard problem that I've been thinking about and discussing with others for some time. It will be an interesting problem to solve.

Reply:
How about a test the automatically and randomly (based on a seed that makes its test case reproducible) creates a test case for one op path every time it runs. It would be a very complicated test because it would need to know, ultimately, how to use every op in TensorFlow. This test could be run at regular intervals, outside of the main CI process (I don't know what you call that; nightly?) and failures could be used to catch op-determinism regressions. It wouldn't automatically catch and block the introduction of the regression, but it would, over time, randomly re-audit every op. That test could be put in place for a small number of ops, to test on all devices, data types, etc and more ops could be added to it over time. It would also need to know details of which op paths/configs would throw d9m-unimplemented exceptions. It might read that from a YAML file. It could keep a record of its progress in a single file somewhere: seed, op, params, result.

Reply:
I've been discussing this with @nluehr. I completely misunderstood the proposal. You want to add this feature under op calls. Then the test suite becomes a ton of pre-generated op calls. This makes sense. Sorry I misunderstood and went off on a tangent.

Okay, so then here are the things I'm worried about:

1. Floats that that have no fractional component. Solution: add/sub a fractional component
2. Calls from the gradient checker where the upstream gradient input is just 1.0. Solution: replace the input with a random number between -0.5 and 0.5.
3. Tensors are too small to cause compute to span asynchronous processors. There doesn't seem to be a simple solution for this problem, and all backward op runs (based on `gradient_checker.compute_gradient`) are going to, necessarily, have tiny tensors.

Reply:
Another class of important stimulus that would likely be missing from existing op tests is where the relationship between input elements specifically triggers the most-common cause of nondeterminism (atomic reduction between asynchronous processes). For example, in segment reduction it's necessary for indices to be repeated more than once (with a large input tensor).

If it was possible to solve the problem of making tensors larger (while not breaking the rules of the op API), then the other half of the problem could be solved by randomizing control inputs (such as segment IDs), but then that would also need to be done in such as way as to not break the rules of the op API.

Reply:
I'm not sure what you originally thought I was proposing, but your issues in the first post make sense (although I didn't understand quite the Jacobian complaint) and your suggestions in the fourth post would fix most of them.

I wonder if we can somehow enlarge the tensors by tiling them on an arbitrary dimension (or trying every dimension in sequence, to make each dimension 10x larger). This is a lot trickier when there are multiple inputs, like Add and MatMul, but we can always catch errors and simply not test that op for determinism with this strategy.

I think it's probably infeasible to for this technique to catch cases where the relationship between input elements cause probable nondeterminism. I doubt we would have caught segment reductions with this technique, but we can try. We can rely on other methods if necessary, like looking for uses of atomics and testing real-world models. 

I think your idea in the third post of having a test suite that reads from a YAML file is reasonable as well, although it requires a small amount of per-op work. Conceptually this is very similar to the approach I suggested, except the configs are hand-written instead of automatically generated based on input shapes and attributes in unit tests. Also in my approach, the determinism tests are run during the unit tests, but perhaps instead the input shapes/attributes/dtypes should be written to a file and afterwards the determinism tests can be run. Then we could augment the file with hand-written tests as well, so that we can give larger inputs that are more likely to demonstrate nondeterminism.